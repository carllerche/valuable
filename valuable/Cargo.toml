[package]
name = "valuable"
version = "0.1.0"
authors = ["Carl Lerche <me@carllerche.com>"]
edition = "2018"
license = "MIT"
description = "Values"

[features]
default = ["std"]

# Provide derive(Valuable) macro
derive = ["valuable-derive"]

# Provide impls for common standard library types like Vec<T> and HashMap<K, V>.
std = ["alloc"]

# Provide imps for types in Rust's `alloc` library.
alloc = []

# Provides serde_json::Value instrument for valuable
json = ["serde_json"]

[dependencies]
valuable-derive = { version = "0.1.0", optional = true, path = "../valuable-derive" }
serde_json = { version = "1.0.46", optional = true }

[dev-dependencies]
criterion = "0.3"

[[bench]]
name = "structable"
harness = false
