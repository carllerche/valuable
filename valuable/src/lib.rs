mod enumerable;
pub use enumerable::{Enumerable, EnumDef, VariantDef};

pub mod field;

mod listable;
pub use listable::Listable;

/*
mod mappable;
pub use mappable::Mappable;
*/

mod record;
pub use record::NamedValues;

mod structable;
pub use structable::{Structable, StructDef};

mod valuable;
pub use valuable::Valuable;

mod value;
pub use value::Value;

mod visit;
pub use visit::Visit;

#[cfg(feature = "derive")]
pub use valuable_derive::*;
