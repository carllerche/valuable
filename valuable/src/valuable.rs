use crate::{Slice, Value, Visit};

use core::fmt;

pub trait Valuable {
    fn as_value(&self) -> Value<'_>;

    fn visit(&self, visit: &mut dyn Visit);

    fn visit_slice(slice: &[Self], visit: &mut dyn Visit)
    where
        Self: Sized,
    {
        const N: usize = 8;

        let mut batch: [_; N] = Default::default();
        let mut curr = 0;

        if slice.is_empty() {
            visit.visit_slice(Slice::Value(&batch[..0]));
            return;
        }

        for v in slice {
            if curr == N {
                visit.visit_slice(Slice::Value(&batch[..]));
                curr = 0;
            }

            batch[curr] = v.as_value();
            curr += 1;
        }

        if curr > 0 {
            visit.visit_slice(Slice::Value(&batch[..curr]));
        }
    }
}

impl<V: ?Sized + Valuable> Valuable for &V {
    fn as_value(&self) -> Value<'_> {
        (*self).as_value()
    }

    fn visit(&self, visit: &mut dyn Visit) {
        V::visit(*self, visit);
    }
}

#[cfg(feature = "alloc")]
impl<V: ?Sized + Valuable> Valuable for alloc::boxed::Box<V> {
    fn as_value(&self) -> Value<'_> {
        (&**self).as_value()
    }

    fn visit(&self, visit: &mut dyn Visit) {
        V::visit(&**self, visit);
    }
}

#[cfg(feature = "alloc")]
impl<V: ?Sized + Valuable> Valuable for alloc::rc::Rc<V> {
    fn as_value(&self) -> Value<'_> {
        (&**self).as_value()
    }

    fn visit(&self, visit: &mut dyn Visit) {
        V::visit(&**self, visit);
    }
}

#[cfg(feature = "alloc")]
impl<V: ?Sized + Valuable> Valuable for alloc::sync::Arc<V> {
    fn as_value(&self) -> Value<'_> {
        (&**self).as_value()
    }

    fn visit(&self, visit: &mut dyn Visit) {
        V::visit(&**self, visit);
    }
}

macro_rules! valuable {
    (
        $(
            $variant:ident($ty:ty),
        )*
    ) => {
        $(
            impl Valuable for $ty {
                fn as_value(&self) -> Value<'_> {
                    Value::$variant(*self)
                }

                fn visit(&self, visit: &mut dyn Visit) {
                    visit.visit_value(self.as_value());
                }

                fn visit_slice(slice: &[Self], visit: &mut dyn Visit)
                where
                    Self: Sized,
                {
                    visit.visit_slice(Slice::$variant(slice));
                }
            }

        )*
    };
}

valuable! {
    Bool(bool),
    Char(char),
    F32(f32),
    F64(f64),
    I8(i8),
    I16(i16),
    I32(i32),
    I64(i64),
    I128(i128),
    Isize(isize),
    U8(u8),
    U16(u16),
    U32(u32),
    U64(u64),
    U128(u128),
    Usize(usize),
}

impl Valuable for () {
    fn as_value(&self) -> Value<'_> {
        Value::Unit
    }

    fn visit(&self, visit: &mut dyn Visit) {
        visit.visit_value(Value::Unit);
    }

    fn visit_slice(slice: &[Self], visit: &mut dyn Visit)
    where
        Self: Sized,
    {
        visit.visit_slice(Slice::Unit(slice));
    }
}

impl Valuable for &'_ str {
    fn as_value(&self) -> Value<'_> {
        Value::String(self)
    }

    fn visit(&self, visit: &mut dyn Visit) {
        visit.visit_value(Value::String(self));
    }

    fn visit_slice(slice: &[Self], visit: &mut dyn Visit)
    where
        Self: Sized,
    {
        visit.visit_slice(Slice::Str(slice));
    }
}

#[cfg(feature = "alloc")]
impl Valuable for alloc::string::String {
    fn as_value(&self) -> Value<'_> {
        Value::String(&self[..])
    }

    fn visit(&self, visit: &mut dyn Visit) {
        visit.visit_value(Value::String(self));
    }

    fn visit_slice(slice: &[Self], visit: &mut dyn Visit)
    where
        Self: Sized,
    {
        visit.visit_slice(Slice::String(slice));
    }
}

#[cfg(feature = "std")]
impl Valuable for dyn std::error::Error + '_ {
    fn as_value(&self) -> Value<'_> {
        Value::Error(self)
    }

    fn visit(&self, visit: &mut dyn Visit) {
        visit.visit_value(self.as_value());
    }
}

impl fmt::Debug for dyn Valuable + '_ {
    fn fmt(&self, fmt: &mut fmt::Formatter<'_>) -> fmt::Result {
        let value = self.as_value();
        value.fmt(fmt)
    }
}
