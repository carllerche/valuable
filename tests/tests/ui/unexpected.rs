use valuable::*;

#[derive(Valuable)]
struct Rename1 {
    #[valuable(rename = b)]
    f: (),
}

#[derive(Valuable)]
struct Transparent1 {
    #[valuable(transparent)]
    f: (),
}
#[derive(Valuable)]
#[valuable(transparent)]
enum Transparent2 {
    V(()),
}
#[derive(Valuable)]
#[valuable(transparent)]
struct Transparent3 {
    f1: (),
    f2: (),
}
#[derive(Valuable)]
#[valuable(transparent, rename = "a")]
struct Transparent4 {
    f: (),
}

#[derive(Valuable)]
#[valuable(skip)]
struct Skip1 {
    f: (),
}
#[derive(Valuable)]
#[valuable(skip)]
enum Skip2 {
    #[valuable(skip)]
    V(()),
}
#[derive(Valuable)]
struct Skip3 {
    #[valuable(skip, rename = "a")]
    f: (),
}

fn main() {}
