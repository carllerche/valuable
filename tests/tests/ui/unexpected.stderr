error: expected `#[valuable(rename = "...")]`, found `#[valuable(rename = ...)]`
 --> tests/ui/unexpected.rs:5:16
  |
5 |     #[valuable(rename = b)]
  |                ^^^^^^^^^^

error: #[valuable(transparent)] may only be used on structs
  --> tests/ui/unexpected.rs:11:16
   |
11 |     #[valuable(transparent)]
   |                ^^^^^^^^^^^

error: #[valuable(transparent)] may only be used on structs
  --> tests/ui/unexpected.rs:15:12
   |
15 | #[valuable(transparent)]
   |            ^^^^^^^^^^^

error: #[valuable(transparent)] struct needs exactly one field, but has 2
  --> tests/ui/unexpected.rs:20:1
   |
20 | / #[valuable(transparent)]
21 | | struct Transparent3 {
22 | |     f1: (),
23 | |     f2: (),
24 | | }
   | |_^

error: #[valuable(transparent)] may not be used together with #[valuable(rename)]
  --> tests/ui/unexpected.rs:26:25
   |
26 | #[valuable(transparent, rename = "a")]
   |                         ^^^^^^

error: #[valuable(skip)] may only be used on fields
  --> tests/ui/unexpected.rs:32:12
   |
32 | #[valuable(skip)]
   |            ^^^^

error: #[valuable(skip)] may only be used on fields
  --> tests/ui/unexpected.rs:37:12
   |
37 | #[valuable(skip)]
   |            ^^^^

error: #[valuable(skip)] may only be used on fields
  --> tests/ui/unexpected.rs:39:16
   |
39 |     #[valuable(skip)]
   |                ^^^^

error: #[valuable(skip)] may not be used together with #[valuable(rename)]
  --> tests/ui/unexpected.rs:44:22
   |
44 |     #[valuable(skip, rename = "a")]
   |                      ^^^^^^
